### ISSUE TITLE
[DRAFT] [Megathread] Removing the sklearn extensions from the openml-python code base

### ISSUE THREAD

**Issue by LennartPurucker** [openml/openml-python#1408]
We are not planning to support the sklearn extensions in the future. 

Currently, the code related to the sklearn extensions within `openml-python` provides a straightforward way to upload models and results from running sklearn models on OpenML. 

Note:
* Even without these extensions, users cannot upload such data to OpenML.
* The sklearn extensions are in the code base of `openml-python`, unlike the code for other extensions.  

## Justification for Removing the Extensions from `openml-python`

As a result of a recent survey by us (to be published) and our interactions with OpenML and its users, we have obtained a strong signal that this feature is not widely used, if at all, by the broader community. Yet, at the same time, it takes up an enormous amount of overhead during maintenance of OpenML with many lines of code, many tests, and multiple reasons for not being able to upgrade dependencies stemming only from the sklearn extension. Moreover, the sklearn extension has not been maintained or upgraded for a long time. 

Previously, a part of the mission statement of OpenML was to enable the reproduction of, for example, old runs via this extension. We do not believe this to be an essential mission statement of the `openml-python` code. More importantly, the extensions are holding us back and are requiring too many of our already limited resources. 

## The Way Forward 

**`openml-python` will remove all code related to the sklearn extension in a future version**

* One will still be able to reproduce the old runs (to some extent) and use the sklearn extension with an older version of `openml-python`
* If there are users of the sklearn extension, or we see renewed need for this extension, a new project can be created --  similar to the other extensions. However, the current `openml-python` team does not commit to supporting this extension. 

These issues will serve as a megathread to reference all code removed related to the sklearn extension and any other content that may enable future readers to recreate the sklearn extension in a separate repository.

**Comment by LennartPurucker** [comment]
A short survey on code we could change/remove as a result: 

- [ ] examples/20_basic/introduction_tutorial.py
- [ ] examples/20_basic/simple_flows_and_runs_tutorial.py (likely replace with code from examples/30_extended/custom_flow_.py)
- [ ] examples/30_extended/fetch_runtimes_tutorial.py
- [ ] examples/30_extended/flow_id_tutorial.py
- [ ] examples/30_extended/flows_and_runs_tutorial.py
- [ ] examples/30_extended/run_setup_tutorial.py
- [ ] examples/30_extended/study_tutorial.py
- [ ] examples/30_extended/task_manual_iteration_tutorial.py (make this the default way to talk about using tasks, etc)
- [ ] examples/40paper (maybe we can remove all of these at the same time?) -> TBD
- [ ] openml/extensions/sklearn
- [ ] LICENSE
- [ ] pyproject.toml (dependencies) 
- [ ] way too many tests (not only for the extension) 
- [ ] CONTRIBUTING.md

**Comment by LennartPurucker** [comment]
Rel.: #1317

**Comment by PGijsbers** [comment]
Tests on `openml-sklearn` are also green [as of now](https://github.com/openml/openml-sklearn/commit/bc9c49cb6d4776536c6f3e6515935afeff2e8417) but it does require installing openml-python from GitHub since otherwise the sklearn extension is registered twice which leads to errors.

**Comment by fkiraly** [comment]
I am a bit late to the discussion, but @LennartPurucker, can you explain how this would make sense? Would this not basically destroy the functionality of the package in python?

My understanding:

* the "core" objects in openml are flows, datasets, tasks, and benchmarks.
* it is not possible to use any of the above without flows
* the sklearn flows are the most basic and most frequently used flows

So ow could removing or deprecating the sklearn flows make sense at all, since you would render the entire package inoperable for python users, through removing its most basic functionality. Maybe I am missing something here.

**Comment by fkiraly** [comment]
also, @LennartPurucker, I find your proposal to deprecate the `openml-sklearn` extension strange.

For instance, you are developing tabarena, a de-facto competitor to the benchmarking functionality in openml.

With your proposal above, this is exactly the kind of benchmarking capability you are removing from openml.

**Comment by LennartPurucker** [comment]
@fkiraly, for your first comment, see https://github.com/openml/openml-python/issues/1532 (not sure in which thread we should continue the discussion to best document it for the public). 

> For instance, you are developing tabarena, a de-facto competitor to the benchmarking functionality in openml.
> With your proposal above, this is exactly the kind of benchmarking capability you are removing from openml.

I would say it is very far-fetched to claim that this functionality is useful or proper benchmarking, especially given that no benchmark has ever used it. As I see it, the extension is a wrapper to upload data to OpenML. In other words, benchmarking is much more than running a user-defined pipeline (and uploading its results to OpenML). I would also add to this that sklearn pipelines (and models) are extremely limited in functionality compared to what researchers and practitioners are using (see TALENT or TabArena models and pipelines for more); hence, the need for functionality that is independent of sklearn. 

The closest we got to having a benchmark that uploads data to OpenML is the [AutoML Benchmark](https://github.com/openml/automlbenchmark) for machine learning systems. But this is technically also independent of this extension and also had some troubles uploading its data back to OpenML (@PGijsbers can expand on this when needed).

**Comment by fkiraly** [comment]
> I would say it is very far-fetched to claim that this functionality is useful or proper benchmarking, especially given that no benchmark has ever used it.

Can you clarify your references, what are you referring to by "it" - tabarena, or openML?

> In other words, benchmarking is much more than running a user-defined pipeline (and uploading its results to OpenML).

How so? Benchmarking is *defined* as evaluating pipelines (or models) on data and then handling/managing/presenting the results.

You seem to want to imply "openml does not do benchmarking properly"? Can you please define what it means to "do benchmarking properly" according to you, and where openml in its design-in-principle falls short (as opposed to its de-facto implementation)?

> I would also add to this that sklearn pipelines (and models) are extremely limited in functionality compared to what researchers and practitioners are using (see TALENT or TabArena models and pipelines for more); hence, the need for functionality that is independent of sklearn.

sklearn pipelines are key baselines that you need to compare to. If you claim you are better than random forest with a fancy tabular foundation model, you actually need to compare to the random forest.

So, in fact, *every state-of-art benchmark for tabular data will need access to the sklearn random forest*, and, a fortiori, to a proper sklearn API.

The argument that we can ignore sklearn because it is (allegedly) not cutting edge does not hold simply by the argument of it being currently still the gold standard for comparing to (not tabPFN etc).

> The closest ...

Even if one would agree that openML currently is insufficient, you dodged a key question:

as an openml core developer, why are you developing "the proper functionality" in TabArena for Amazon, and not for openML, if you think it is not good enough?

**Comment by LennartPurucker** [comment]
> Can you clarify your references, what are you referring to by "it" - tabarena, or openML?

`openml-python-sklearn`

> How so? Benchmarking is defined as evaluating pipelines (or models) on data and then handling/managing/presenting the results. You seem to want to imply "openml does not do benchmarking properly"? Can you please define what it means to "do benchmarking properly" according to you, and where openml in its design-in-principle falls short (as opposed to its de-facto implementation)?


Please read the TabArena paper, or any other tabular benchmark paper, to get a better sense of what constitutes benchmarking. And then, compare this to the functionality offered by `openml-python-sklearn`. 

Also, we should clarify the semantics; OpenML does not "do" anything. It is a platform that enables people to "do" things. 


> sklearn pipelines are key baselines that you need to compare to. If you claim you are better than random forest with a fancy tabular foundation model, you actually need to compare to the random forest. So, in fact, every state-of-art benchmark for tabular data will need access to the sklearn random forest, and, a fortiori, to a proper sklearn API. The argument that we can ignore sklearn because it is (allegedly) not cutting edge does not hold simply by the argument of it being currently still the gold standard for comparing to (not tabPFN etc).

You are coming into this argument very strongly, while you seem to lack substantial background knowledge. I am happy to talk about this more, but this is clearly the wrong place to do so. 

To correct some of your false impressions shortly: (1) the taublar benchmarking community has constantly and is still comparing to sklearn pipelines (including TabArena) and it seems pretty trivial to be better than these pipelines in all major past and present benchmarks, and (2) I never said we should ignore the models from sklearn, I said " sklearn pipelines (and models) are extremely limited in functionality", which means they don't support a lot of things required of them in practice, which other models used by practitioners do (such as XGBoost, CatBoost, LGBM).

> as an openml core developer, why are you developing "the proper functionality" in TabArena for Amazon, and not for openML, if you think it is not good enough?

Could you elaborateon this statement, please: "TabArena for Amazon" 

Do you have any problems with the open-source projects that is TabArena? What am I missing that makes you think TabArena is "for Amazon"? 

And to answer your problematically framed question: the functionality we developed for TabArena (such as models, pipelines, datasets, ...) is clearly out of scope for OpenML as a platform. Just as the AutoML benchmark is not part of `openml-python-sklearn`.

**Comment by fkiraly** [comment]
> I never said we should ignore the models from sklearn, I said " sklearn pipelines (and models) are extremely limited in functionality"

You said this in the top post: "We are not planning to support the sklearn extensions in the future."

That sounds exactly like saying we should ignore them (in the purview of openml at least).

> Please read the TabArena paper, or any other tabular benchmark paper, to get a better sense of what constitutes benchmarking. 

This is a pretty nasty and aggressive way to dodge my perfectly reasonable question: a claim of lack of knowledge, and deflecting by requesting me to read a paper without saying specifically what I should be looking for.

I am not going to re-read the paper, because the problem is not my knowledge but it was a question about your opinion.

I want to hear your specific response to my question.

So, can you please define what it means to "do benchmarking properly" according to you, and where openml in its design-in-principle falls short (as opposed to its de-facto implementation)?

This is a genuine, and not a rhetorical question, directed at you. I expect constructive discussion, thanks.

> Could you elaborateon this statement, please: "TabArena for Amazon"

The autogluon org and brand are legally controlled by Amazon US, it is an Amazon brand.

On the landing page: "AutoGluon, developed by AWS AI", as of today.
https://github.com/autogluon/autogluon

TabArena is in the AutoGluon organisation, hence also legally controlled by Amazon.

> the functionality we developed for TabArena (such as models, pipelines, datasets, ...) is clearly out of scope for OpenML as a platform.

Curious, I thought this is exactly in-scope for openml? Where pipelines are called "flows".

If you disagree, can you please elaborate how the TabArena scope differs from OpenML? Here, please assume the scope as in the original OpenML papers, not your personal preference of what the roadmap *should* be.

**Comment by LennartPurucker** [comment]
> You said this in the top post: "We are not planning to support the sklearn extensions in the future." That sounds exactly like saying we should ignore them (in the purview of openml at least).

Us not supporting something vs. us ignoring something are quite different things, no? :) 

> The autogluon org and brand are legally controlled by Amazon US, it is an Amazon brand.

Could you please provide some reference or proof of this statement? Especially that AutoGluon is a brand or legally controlled by Amazon? I cannot find a brand or trademark registry if it exists. 

Even if it exists, this should not stop us from working with great open-source projects, as none of this affects their licenses or our ability to create a fork under a new brand. Both projects are open-source: 

- https://github.com/autogluon/autogluon/blob/master/LICENSE
- https://github.com/autogluon/tabarena/blob/main/LICENSE

I guess you're saying that AutoGluon and TabArena are "controlled" by the people maintaining these projects. AutoGluon, an open-source project, is primarily developed and led by AWS AI, with many external contributors: https://github.com/autogluon/autogluon/graphs/contributors. TabArena has more external than internal developers, and it is an active collaboration we hosted under the AutoGluon organization. If there were good reasons to change organizations, we could do that (such as our HF organization https://huggingface.co/TabArena). But TabArena's ecosystem is tightly coupled to the AutoGluon ecosystem. So it is a good match. 

On this note, I would love to know your opinion on probabl, ["the official and exclusive operator of scikit-learn" and its VC funding](https://papers.probabl.ai/probabl-raises-a-13m-in-seed-to-accelerate-enterprise-grade-ai)! 

---

>  "do benchmarking properly" according to you

In short, proper benchmarking is benchmarking without any methodological mistakes. 

There is a long list of methodological mistakes that occurred in the past. We wrote an entire paper about this, included all relevant related work on "proper benchmarking", and a very long appendix with more information for the interested reader. So I won't reiterate it here as it is not worth my time (or read as "I am too lazy" if you want to), and instead, again, point you to the literature that you can read on your own time: https://arxiv.org/abs/2506.16791 :) 

>  and where openml in its design-in-principle falls short (as opposed to its de-facto implementation)?

OpenML does not design-in-principle a benchmark, but a way to share benchmark artifacts. 


> Curious, I thought this is exactly in-scope for openml? Where pipelines are called "flows".

So you would want to implement LightGBM with early-stopping cross-validation and a bagging ensemble in `openml-python-sklearn`? 

Again, semantics are essential: we want support for uploading the details of any LightGBM pipeline, but we don't want to support the pipeline's code itself (e.g., we don't develop sklearn's random forest). 

---

We can discuss the rest of your points on Slack / the meeting.
I believe that either format fits this discussion better than these GitHub issues.

**Comment by fkiraly** [comment]
> Us not supporting something vs. us ignoring something are quite different things, no? :)

No, these are both ways to kill something.
(if a difference exists at all, it is in the degree of plausible deniability)

> Could you please provide some reference or proof of this statement? Especially that AutoGluon is a brand or legally controlled by Amazon?

I already provided above. It says that autogluon is an Amazon project [right at the top of the readme](https://github.com/autogluon/autogluon):

<img width="566" height="77" alt="Image" src="https://github.com/user-attachments/assets/1e82cdda-2633-4a36-af20-8fbee091ff43" />

The project owner(s) on GitHub and pypi are Amazon staff:

<img width="730" height="639" alt="Image" src="https://github.com/user-attachments/assets/f1211963-13b8-4ab8-83d9-9219bb5c8f87" />

In the legal sphere, this would typically be sufficient to argue that the project is owned/controlled by Amazon, since it controls the key tech assets (GitHub repo and pypi distribution pathway).

> So I won't reiterate it here as it is not worth my time (or read as "I am too lazy" if you want to), and instead, again, point you to the literature that you can read on your own time: https://arxiv.org/abs/2506.16791 

This statement feels very egocentric and blasé, you are basically saying anyone who was not read your paper on benchmarking does not understand the topic. This is furthermore one on "tabarena", not on benchmarking in general.

I have also written papers on tabular benchmarking, which you have apparently neither read nor cited. I will not return the "favour" of accusing you of incompetence. But, last time I looked, it was proper academic practice to cite relevant work. You can read my papers here:

* https://arxiv.org/abs/1812.07519
* https://arxiv.org/abs/1901.03678

(one of these is even on an experimental python toolbox design for tabular benchmarking)

**Comment by fkiraly** [comment]
> So you would want to implement LightGBM with early-stopping cross-validation and a bagging ensemble in openml-python-sklearn?

Yes. Wouldn't you?

Scientifically: LightGBM (and xgboost) are key "state-of-the-art baselines" that ought to be included, and they are, in terms of type, sklearn-like.

Architecture-wise: it is better to use the sklearn API to these that already exists than writing new connectors or adapters.
